framework/src/main/java/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -2568,34 +2540,50 @@ public AnnotatedTypeMirror getDefaultValueAnnotatedType(TypeMirror typeMirror) {
 
   /**
    * Returns a {@code RequiresQualifier("...")} or {@code EnsuresQualifier("...")} annotation for
-   * the given field. Returns null if none can be created, because the qualifier has
+   * the given expression. Returns null if none can be created, because the qualifier has
    * elements/arguments, which {@code @RequiresQualifier} and {@code @EnsuresQualifier} do not
-   * support.
+   * support. Also returns null if the expression is invalid when combined with the kind of
+   * annotation: for example, precondition annotations on "this" and parameters ("#1", etc.) are not
+   * supported, because receiver/parameter annotations should be inferred instead.
    *
    * <p>This is of the form {@code @RequiresQualifier(expression="this.elt",
    * qualifier=MyQual.class)} or {@code @EnsuresQualifier(expression="this.elt",
    * qualifier=MyQual.class)} when elt is declared as {@code @A} or {@code @Poly*} and f contains
    * {@code @B} which is a sub-qualifier of {@code @A}.
    *
-   * @param fieldElement a field
+   * @param expression the expression to which the qualifier applies
    * @param qualifier the qualifier that must be present
+   * @param declaredType the declared type of the expression, which is used to avoid inferring
+   *     redundant pre- or postcondition annotations
    * @param preOrPost whether to return a precondition or postcondition annotation
+   * @param preconds the list of precondition annotations; always null if {@code preOrPost} is
+   *     {@code BeforeOrAfter.BEFORE}
    * @return a {@code RequiresQualifier("...")} or {@code EnsuresQualifier("...")} annotation for
    *     the given field, or null
    */
-  protected @Nullable AnnotationMirror requiresOrEnsuresQualifierAnno(
-      VariableElement fieldElement, AnnotationMirror qualifier, Analysis.BeforeOrAfter preOrPost) {
+  protected @Nullable AnnotationMirror requiresOrEnsuresQualifierAnnoForExpression(
+      String expression,
+      AnnotationMirror qualifier,
+      AnnotatedTypeMirror declaredType,
+      Analysis.BeforeOrAfter preOrPost,
+      @Nullable List<AnnotationMirror> preconds) {
+
+    // Do not generate RequiresQualifier annotations for "this" or parameter expressions.
+    if (preOrPost == BeforeOrAfter.BEFORE
+        && ("this".equals(expression) || expression.startsWith("#"))) {
```

Comment:
`expression.startsWith("#")` is too lenient because it also matches, for example, `#1.f`.